File: Retos/Reto #6 - PIEDRA, PAPEL, TIJERA, LAGARTO, SPOCK [Media]/php/lombervid.php
<?php

declare(strict_types=1);

enum Options: string
{
    case Rock = '🗿';
    case Paper = '📄';
    case Scissors = '✂️';
    case Lizard = '🦎';
    case Spock = '🖖';

    protected function priority(): array
    {
        return match ($this) {
            self::Rock => [self::Lizard, self::Scissors],
            self::Paper => [self::Rock, self::Spock],
            self::Scissors => [self::Paper, self::Lizard],
            self::Lizard => [self::Spock, self::Paper],
            self::Spock => [self::Scissors, self::Rock],
            default => [],
        };
    }

    public function wins(Options $option): bool
    {
        return in_array($option, $this->priority());
    }

    public static function toString()
    {
        return implode(' | ', array_column(self::cases(), 'value'));
    }
}

class GameOptions implements IteratorAggregate
{
    protected array $options = [];

    public function __construct(array $options = null)
    {
        if ($options) {
            $this->setOptions($options);
        }
    }

    public function setOptions(array $options)
    {
        $newOptions = [];

        foreach ($options as $key => $option) {
            if (!is_array($option)) {
                throw new InvalidArgumentException("Every option must be an array on key '{$key}'");
            }

            try {
                $newOptions[] = [Options::from($option[0] ?? null), Options::from($option[1] ?? null)];
            } catch (Throwable $t) {
                throw new InvalidArgumentException("Options must be of type (at index {$key}): " . Options::toString());
            }
        }

        $this->options = $newOptions;
    }

    public function getIterator(): Traversable
    {
        return new ArrayIterator($this->options);
    }
}

class LizardSpockGame
{
    protected int $score1 = 0;
    protected int $score2 = 0;

    public function __construct(
        protected GameOptions $options = new GameOptions(),
    ) {
    }

    public function setInput(array $options)
    {
        $this->options->setOptions($options);

        return $this;
    }

    public function play(): string
    {
        $this->resetScores();

        foreach ($this->options as [$p1, $p2]) {
            if ($p1->wins($p2)) {
                $this->score1 += 1;
            } elseif ($p2->wins($p1)) {
                $this->score2 += 1;
            }
        }

        if ($this->score1 === $this->score2) {
            return 'Tie';
        }

        return ($this->score1 > $this->score2) ? 'Player 1' : 'Player 2';
    }

    protected function resetScores()
    {
        $this->score1 = 0;
        $this->score2 = 0;
    }
}

/**
 * Test cases
 *
 * Rock     = '🗿'
 * Paper    = '📄'
 * Scissors = '✂️'
 * Lizard   = '🦎'
 * Spock    = '🖖'
 *
 */

$failures = 0;
$tests = [
    [
        [['🗿','✂️'], ['✂️','🗿'], ['📄','✂️']],
        'Player 2',
    ],
    [
        [['🖖','✂️'], ['📄','🗿'], ['✂️', '🦎'], ['🖖', '🦎'], ['📄', '🗿']],
        'Player 1',
    ],
    [
        [['🖖', '🖖']],
        'Tie',
    ],
    [
        [['🖖', '🗿'], ['🦎', '✂️'], ['📄', '🖖'], ['🦎', '🗿']],
        'Tie',
    ],
];

$game = new LizardSpockGame();

foreach ($tests as $index => [$input, $want]) {
    try {
        $got = $game->setInput($input)->play();

        if ($got !== $want) {
            $failures += 1;
            echo "Test {$index} failed: got \"{$got}\", want \"{$want}\"" . PHP_EOL;
        } else {
            echo $got . PHP_EOL;
        }
    } catch (Throwable $t) {
        $failures += 1;
        echo "Test {$index} failed with message: " . $t->getMessage() . PHP_EOL;
    }
}

if ($failures === 0) {
    echo "All tests complete successfuly" . PHP_EOL;
}

File: Retos/Reto #6 - PIEDRA, PAPEL, TIJERA, LAGARTO, SPOCK [Media]/go/lombervid.go
package main

import (
	"errors"
	"fmt"
)

type Option string

const (
	Rock     Option = "🗿"
	Paper    Option = "📄"
	Scissors Option = "✂️"
	Lizard   Option = "🦎"
	Spock    Option = "🖖"
)

type Strongness [2]Option
type Options map[Option]Strongness

var options = Options{
	Rock:     {Lizard, Scissors},
	Paper:    {Rock, Spock},
	Scissors: {Paper, Lizard},
	Lizard:   {Spock, Paper},
	Spock:    {Scissors, Rock},
}

func (wn Strongness) Contains(symbol Option) bool {
	for _, opt := range wn {
		if opt == symbol {
			return true
		}
	}
	return false
}

type GameOption struct {
	symbol Option
	winsTo Strongness
}

func (gopt GameOption) Wins(option GameOption) bool {
	return gopt.winsTo.Contains(option.symbol)
}

func NewGameOption(symbol Option) (*GameOption, error) {
	if winsTo, ok := options[symbol]; ok {
		return &GameOption{symbol, winsTo}, nil
	}
	return nil, errors.New("Invalid Option")
}

type GameMatch struct {
	p1 *GameOption
	p2 *GameOption
}

type LizardSpockGame struct {
	score1  int
	score2  int
	options []GameMatch
}

func (lsg *LizardSpockGame) resetScores() {
	lsg.score1 = 0
	lsg.score2 = 0
}

func (lsg *LizardSpockGame) SetInputs(inputs [][2]Option) *LizardSpockGame {
	var options []GameMatch

	for i, match := range inputs {
		opt1, err := NewGameOption(match[0])
		if err != nil {
			panic(fmt.Sprintf("%s (at index %d:0)", err, i))
		}

		opt2, err := NewGameOption(match[1])
		if err != nil {
			panic(fmt.Sprintf("%s (at index %d:1)", err, i))
		}

		options = append(options, GameMatch{opt1, opt2})
	}

	lsg.options = options

	return lsg
}

func (lsg *LizardSpockGame) Play() string {
	lsg.resetScores()

	for _, option := range lsg.options {
		if option.p1.Wins(*option.p2) {
			lsg.score1++
		}
		if option.p2.Wins(*option.p1) {
			lsg.score2++
		}
	}

	if lsg.score1 > lsg.score2 {
		return "Player 1"
	}

	if lsg.score2 > lsg.score1 {
		return "Player 2"
	}

	return "Tie"
}

func main() {
	var failures int
	tests := []struct {
		input [][2]Option
		want  string
	}{
		{
			[][2]Option{{"🗿", "✂️"}, {"✂️", "🗿"}, {"📄", "✂️"}},
			"Player 2",
		},
		{
			[][2]Option{{"🖖", "✂️"}, {"📄", "🗿"}, {"✂️", "🦎"}, {"🖖", "🦎"}, {"📄", "🗿"}},
			"Player 1",
		},
		{
			[][2]Option{{"🖖", "🖖"}},
			"Tie",
		},
		{
			[][2]Option{{"🖖", "🗿"}, {"🦎", "✂️"}, {"📄", "🖖"}, {"🦎", "🗿"}},
			"Tie",
		},
	}

	game := &LizardSpockGame{}

	for i, test := range tests {
		func() {
			defer func() {
				if r := recover(); r != nil {
					failures++
					fmt.Printf("Test %d failed with message: %q\n", i, r)
				}
			}()

			got := game.SetInputs(test.input).Play()

			if got != test.want {
				failures++
				fmt.Printf("Test %d failed: got %q, want %q\n", i, got, test.want)
				return
			}

			fmt.Println(got)
		}()
	}

	if failures == 0 {
		fmt.Println("All tests passed")
	}
}
